### PR TITLE
Issues/16 data sharing

### LINKED ISSUE #16: Create a data-sharing process for the Projectathon 7
The process could look something like this:

Datenbereitstellung2

### LINKED ISSUE #20: Use SignalEvent to remove old KDS Autostart Process Instances
Currently, the KDS Autostart report uses an implementation to search for existing process instances and stops them, before starting a new one. This has the disadvantage, that Task resources of stopped process instances remain in the state `in-progress`.

Using a BPMN Signal Event as broadcaster, the current implementation can be removed and existing process instances can be terminated correctly using the existing default EndEventListener.

### PR DESCRIPTION
**Pull request by wetret**
- Creates a basic data-sharing process for the projectathon 7
- improves the autostart KDS report process by using a signal event

resolves #16
resolves #20

### CODE REVIEW

File: mii-dsf-process-projectathon-data-sharing/src/main/resources/fhir/StructureDefinition/mii-projectathon-task-merge-data-sharing.xml [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+      </slicing>
+    </element>
+    <element id="Task.input:medic-correlation-key.extension:medic-identifier">
+      <path value="Task.input.extension" />
+      <sliceName value="medic-identifier" />
+      <min value="1" />
+      <max value="1" />
+      <type>
+        <code value="Extension" />
+        <profile value="http://medizininformatik-initiative.de/fhir/StructureDefinition/extension-medic-identifier" />
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/resources/fhir/StructureDefinition/mii-projectathon-task-merge-data-sharing.xml:241]: Use profile with version, if the extension is defined in the process plugin
```suggestion
        <profile value="http://medizininformatik-initiative.de/fhir/StructureDefinition/extension-medic-identifier|#{version}" />
```

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/merge/CommunicateMissingDataSetsMerge.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+				target.getOrganizationIdentifierValue(), projectIdentifier, taskId);
+	}
+
+	private void sendMail(Targets targets, String projectIdentifier)
+	{
+		String subject = "Missing data-sets in process '" + ConstantsDataSharing.PROCESS_NAME_FULL_MERGE_DATA_SHARING
+				+ "'";
+		StringBuilder message = new StringBuilder("Data-sets are missing for data-sharing project '" + projectIdentifier
+				+ "' in process '" + ConstantsDataSharing.PROCESS_NAME_FULL_MERGE_DATA_SHARING
+				+ "' from the following organizations:\n");
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/merge/CommunicateMissingDataSetsMerge.java:74]: If you are using a StringBuilder for performance reasons, use `append` for all the Strings.

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/StoreDataSet.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+					+ "' and  data-sharing project '" + projectIdentifier + "' referenced in Task with id '"
+					+ task.getId() + "' - " + exception.getMessage();
+
+			execution.setVariable(ConstantsDataSharing.BPMN_EXECUTION_VARIABLE_DATA_SHARING_EXECUTE_ERROR_MESSAGE,
+					Variables.stringValue(message));
+
+			throw new BpmnError(ConstantsDataSharing.BPMN_EXECUTION_VARIABLE_DATA_SHARING_EXECUTE_ERROR, message,
+					exception);
+		}
+
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/StoreDataSet.java:92]: Empty line.

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/SelectDataSetTarget.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import org.highmed.dsf.fhir.variables.Target;
+import org.highmed.dsf.fhir.variables.TargetValues;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.medizininformatik_initiative.process.projectathon.data_sharing.ConstantsDataSharing;
+
+public class SelectDataSetTarget extends AbstractServiceDelegate
+{
+	private static final Logger logger = LoggerFactory.getLogger(SelectDataSetTarget.class);
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/SelectDataSetTarget.java:22]: Logger is never used

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/coordinate/SelectCosTarget.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import org.highmed.dsf.fhir.variables.TargetValues;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.InitializingBean;
+
+import de.medizininformatik_initiative.process.projectathon.data_sharing.ConstantsDataSharing;
+
+public class SelectCosTarget extends AbstractServiceDelegate implements InitializingBean
+{
+	private static final Logger logger = LoggerFactory.getLogger(SelectCosTarget.class);
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/coordinate/SelectCosTarget.java:23]: Logger is never used

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/HandleErrorExecute.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+	@Override
+	protected void doExecute(DelegateExecution execution)
+	{
+		String error = (String) execution
+				.getVariable(ConstantsDataSharing.BPMN_EXECUTION_VARIABLE_DATA_SHARING_EXECUTE_ERROR_MESSAGE);
+		String projectIdentifier = (String) execution
+				.getVariable(ConstantsDataSharing.BPMN_EXECUTION_VARIABLE_PROJECT_IDENTIFIER);
+
+		logger.warn(error + " - creating new user-task 'release-data-set'");
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/HandleErrorExecute.java:45]:
```suggestion
		logger.warn("{} - creating new user-task 'release-data-set'", error);
```

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/PrepareExecution.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	private String getProjectIdentifier(Task task)
+	{
+		return task.getInput().stream().filter(i -> i.getType().getCoding().stream()
+				.anyMatch(c -> ConstantsDataSharing.CODESYSTEM_DATA_SHARING.equals(c.getSystem())
+						&& ConstantsDataSharing.CODESYSTEM_DATA_SHARING_VALUE_PROJECT_IDENTIFIER.equals(c.getCode())))
+				.filter(i -> i.getValue() instanceof Identifier).map(i -> (Identifier) i.getValue())
+				.filter(i -> ConstantsDataSharing.NAMINGSYSTEM_PROJECT_IDENTIFIER.equals(i.getSystem()))
+				.map(Identifier::getValue).findFirst().orElseThrow(() -> new RuntimeException(
+						"No project-identifier present in task with id='" + task.getId() + "'"));
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/PrepareExecution.java:58]:
```suggestion
						"No project-identifier present in task with id '" + task.getId() + "'"));
```

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/PrepareExecution.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	private String getCosIdentifier(Task task)
+	{
+		return getTaskHelper()
+				.getInputParameterReferenceValues(task, ConstantsDataSharing.CODESYSTEM_DATA_SHARING,
+						ConstantsDataSharing.CODESYSTEM_DATA_SHARING_VALUE_COS_IDENTIFIER)
+				.filter(Reference::hasIdentifier).map(Reference::getIdentifier).map(Identifier::getValue).findFirst()
+				.orElseThrow(
+						() -> new RuntimeException("No COS-identifier found in Task with id='" + task.getId() + "'"));
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/PrepareExecution.java:68]:
```suggestion
						() -> new RuntimeException("No COS-identifier found in Task with id '" + task.getId() + "'"));
```

File: mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/PrepareExecution.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	private String getContractLocation(Task task)
+	{
+		return getTaskHelper()
+				.getFirstInputParameterUrlValue(task, ConstantsDataSharing.CODESYSTEM_DATA_SHARING,
+						ConstantsDataSharing.CODESYSTEM_DATA_SHARING_VALUE_CONTRACT_LOCATION)
+				.map(UrlType::getValue).orElseThrow(() -> new RuntimeException(
+						"No project-identifier present in task with id='" + task.getId() + "'"));
```
Comment (hhund) [inline mii-dsf-process-projectathon-data-sharing/src/main/java/de/medizininformatik_initiative/process/projectathon/data_sharing/service/execute/PrepareExecution.java:77]:
```suggestion
						"No project-identifier present in task with id '" + task.getId() + "'"));
```

**Review by hhund — APPROVED** [review]
I found a few minor things you may want to change.

File: mii-dsf-process-projectathon-data-sharing/src/main/resources/bpe/coordinate.bpmn [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      <bpmn:incoming>Flow_00ok4yz</bpmn:incoming>
+      <bpmn:outgoing>Flow_1gu8xzh</bpmn:outgoing>
+    </bpmn:serviceTask>
+    <bpmn:sequenceFlow id="Flow_1d6fc4e" sourceRef="extractMergedDataSetLocation" targetRef="coordinateDataSharingEndEvent" />
+    <bpmn:serviceTask id="extractMergedDataSetLocation" name="extract merged data-set location" camunda:class="de.medizininformatik_initiative.process.projectathon.data_sharing.service.coordinate.ExtractMergedDataSetLocation">
+      <bpmn:incoming>Flow_1gu8xzh</bpmn:incoming>
+      <bpmn:outgoing>Flow_1d6fc4e</bpmn:outgoing>
+    </bpmn:serviceTask>
+  </bpmn:process>
+  <bpmn:message id="Message_2tdo6h2" name="coordinateDataSharing" />
```
[the remark this answers is not in the document]
Comment (wetret) [reply mii-dsf-process-projectathon-data-sharing/src/main/resources/bpe/coordinate.bpmn:88]: Thats a good point! We should create a guideline for process naming conventions not only in message names and then adapt all processes at once.

**Comment by wetret** [comment]
Closing without merge because of wrong base branch `main` instead of `develop`. Commits have been merged into `develop`.

**Review by hhund — APPROVED** [review]
Logger in SelectDicTargets is never used and can be removed. As discussed elsewhere the HRP organization should execute a Ping/Pong process, to insure all organizations are accessible before starting the `coordinate` process. Otherwise this looks like a good implementation for the 7. MII Projectathon.

**Pull request merged**